### ISSUE TITLE
Network layer order exception

### ISSUE THREAD

**Issue by tvanh** [deeplearning4j/deeplearning4j#6613]
I tried to mimic a Super Resolution Image network

The network conf is here

`
	public ComputationGraphConfiguration EEDS(){
        ComputationGraphConfiguration.GraphBuilder graph = new NeuralNetConfiguration.Builder().seed(seed)
                .optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT)
                .updater(updater)
                .weightInit(weightInit)
                .dist(new TruncatedNormalDistribution(0.0, 0.5))
                .l2(5e-5)
                .miniBatch(true)
                .cacheMode(cacheMode)
                .trainingWorkspaceMode(workspaceMode)
                .inferenceWorkspaceMode(workspaceMode)
                .graphBuilder();
        
        return 
           graph.addInputs("input").setInputTypes(InputType.convolutional(h, w, 3))
        
        		//feature extraction
        		.addLayer("fe-conv-1", new ConvolutionLayer.Builder(3,3).stride(1,1).nOut(64)
		                .convolutionMode(ConvolutionMode.Same).cudnnAlgoMode(cudnnAlgoMode)
		                .activation(Activation.RELU).build(), "input")
        		.addLayer("fe-conv-2", new ConvolutionLayer.Builder(3,3).stride(1,1).nOut(64)
		                .convolutionMode(ConvolutionMode.Same).cudnnAlgoMode(cudnnAlgoMode)
		                .activation(Activation.RELU).build(), "fe-conv-1")
        		.addLayer("fe-conv-3", new ConvolutionLayer.Builder(3,3).stride(1,1).nOut(64)
		                .convolutionMode(ConvolutionMode.Same).cudnnAlgoMode(cudnnAlgoMode)
		                .activation(Activation.IDENTITY).build(), "fe-conv-2")
        		.addVertex("fe-add-vertex", new ElementWiseVertex(Op.Add), "fe-conv-1", "fe-conv-3") // add layer 1 and 3
        		.addLayer("fe-relu", new ActivationLayer.Builder().activation(Activation.RELU).build(), "fe-add-vertex")
        		//Upsampling 
        		.addLayer("up-conv-1", new ConvolutionLayer.Builder(1,1).stride(1,1).nOut(4)
		                .convolutionMode(ConvolutionMode.Same).cudnnAlgoMode(cudnnAlgoMode)
		                .activation(Activation.RELU).build(), "fe-relu")
        		.addLayer("up-deconv-1", new Deconvolution2D.Builder(15, 15).stride(1, 1).nOut(4)
		                .convolutionMode(ConvolutionMode.Strict).cudnnAlgoMode(cudnnAlgoMode)
		                .activation(Activation.RELU).build(), "up-conv-1")
        		.addLayer("up-conv-2", new ConvolutionLayer.Builder(1,1).stride(1,1).nOut(64)
		                .convolutionMode(ConvolutionMode.Same).cudnnAlgoMode(cudnnAlgoMode)
		                .activation(Activation.RELU).build(), "up-deconv-1")
        		//Multi-scale Reconstruction
        		//recontruct-block-1
        		.addLayer("recon-conv-1-1", new ConvolutionLayer.Builder(3,3).stride(1,1).nOut(64)
		                .convolutionMode(ConvolutionMode.Same).cudnnAlgoMode(cudnnAlgoMode)
		                .activation(Activation.RELU).build(), "up-conv-2")
        		.addLayer("recon-conv-1-2", new ConvolutionLayer.Builder(3,3).stride(1,1).nOut(64)
		                .convolutionMode(ConvolutionMode.Same).cudnnAlgoMode(cudnnAlgoMode)
		                .activation(Activation.IDENTITY).build(), "recon-conv-1-1")
        		.addVertex("recon-add-vertex-1", new ElementWiseVertex(Op.Add), "up-conv-2", "recon-conv-1-2") // add layer 1 and 3
        		.addLayer("recon-relu-1", new ActivationLayer.Builder().activation(Activation.RELU).build(), "recon-add-vertex-1")
        		//recontruct-block-2
        		.addLayer("recon-conv-2-1", new ConvolutionLayer.Builder(3,3).stride(1,1).nOut(64)
		                .convolutionMode(ConvolutionMode.Same).cudnnAlgoMode(cudnnAlgoMode)
		                .activation(Activation.RELU).build(), "recon-relu-1")
        		.addLayer("recon-conv-2-2", new ConvolutionLayer.Builder(3,3).stride(1,1).nOut(64)
		                .convolutionMode(ConvolutionMode.Same).cudnnAlgoMode(cudnnAlgoMode)
		                .activation(Activation.IDENTITY).build(), "recon-conv-2-1")
        		.addVertex("recon-add-vertex-2", new ElementWiseVertex(Op.Add), "recon-relu-1", "recon-conv-2-2") // add layer 1 and 3
        		.addLayer("recon-relu-2", new ActivationLayer.Builder().activation(Activation.RELU).build(), "recon-add-vertex-2")
        		//multi-scale
        		.addLayer("scale-conv-1", new ConvolutionLayer.Builder(1,1).stride(1,1).nOut(16)
		                .convolutionMode(ConvolutionMode.Same).cudnnAlgoMode(cudnnAlgoMode)
		                .activation(Activation.RELU).build(), "recon-relu-2")
        		.addLayer("scale-conv-2-1", new ConvolutionLayer.Builder(1,1).stride(1,1).nOut(16)
		                .convolutionMode(ConvolutionMode.Same).cudnnAlgoMode(cudnnAlgoMode)
		                .activation(Activation.RELU).build(), "scale-conv-1")
        		.addLayer("scale-conv-2-2", new ConvolutionLayer.Builder(3,3).stride(1,1).nOut(16)
		                .convolutionMode(ConvolutionMode.Same).cudnnAlgoMode(cudnnAlgoMode)
		                .activation(Activation.RELU).build(), "scale-conv-2-1")
        		.addLayer("scale-conv-2-3", new ConvolutionLayer.Builder(5,5).stride(1,1).nOut(16)
		                .convolutionMode(ConvolutionMode.Same).cudnnAlgoMode(cudnnAlgoMode)
		                .activation(Activation.RELU).build(), "scale-conv-2-2")
        		.addLayer("scale-conv-2-4", new ConvolutionLayer.Builder(7,7).stride(1,1).nOut(16)
		                .convolutionMode(ConvolutionMode.Same).cudnnAlgoMode(cudnnAlgoMode)
		                .activation(Activation.RELU).build(), "scale-conv-2-3")
        		.addLayer("scale-conv-3", new ConvolutionLayer.Builder(1,1).stride(1,1).nOut(1)
		                .convolutionMode(ConvolutionMode.Same).cudnnAlgoMode(cudnnAlgoMode)
		                .activation(Activation.RELU).build(), "scale-conv-2-4")
        		
        		//Shallow network
        		.addLayer("shallow-1", new ConvolutionLayer.Builder(3,3).stride(1,1).nOut(4)
		                .convolutionMode(ConvolutionMode.Same).cudnnAlgoMode(cudnnAlgoMode)
		                .activation(Activation.RELU).build(), "input")
        		.addLayer("shallow-2", new Deconvolution2D.Builder(15, 15).stride(1, 1).nOut(8)
		                .convolutionMode(ConvolutionMode.Strict).cudnnAlgoMode(cudnnAlgoMode)
		            ```
    .activation(Activation.RELU).build(), "shallow-1")
        		.addLayer("shallow-3", new ConvolutionLayer.Builder(1,1).stride(1,1).nOut(1)
		                .convolutionMode(ConvolutionMode.Same).cudnnAlgoMode(cudnnAlgoMode)
		                .activation(Activation.RELU).build(), "shallow-2")

        		//combine
        		.addVertex("combine", new ElementWiseVertex(Op.Add), "scale-conv-3", "shallow-3") // combine deep and shallow
        		//output
        		.addLayer("output", new CnnLossLayer.Builder(LossFunctions.LossFunction.MCXENT).build(), "combine")
        		.setOutputs("output")
        		//build
        		.build();
	}
`
The data is load using this 

`	
	public DataSet next(int num) {
		try {
			File file = files.get(cursor);
			cursor++;
			
	        CropImageTransform crop = new CropImageTransform(h, w, h, w);
	        ScaleImageTransform scale = new ScaleImageTransform(w/2, h/2);
	        		
	        
	        NativeImageLoader imageLoader = new NativeImageLoader(0, 0, 3);
	        ImageWritable wi;
				wi = imageLoader.asWritable(file);
	        //sr
	        ImageWritable srImage = crop.transform(wi);
	        //lr
	        ImageWritable lrImage = scale.transform(srImage);
	        INDArray input = imageLoader.asMatrix(lrImage);
	        INDArray output = imageLoader.asMatrix(srImage);
	        
	    	return new DataSet(input, output);	
		} catch (IOException e) {
			e.printStackTrace();
			return null;
		}
    }
`

And I got this exception

Exception in thread "main" java.lang.NullPointerException
	at org.deeplearning4j.nn.updater.graph.ComputationGraphUpdater.getOrderedLayers(ComputationGraphUpdater.java:63)
	at org.deeplearning4j.nn.updater.BaseMultiLayerUpdater.<init>(BaseMultiLayerUpdater.java:75)
	at org.deeplearning4j.nn.updater.graph.ComputationGraphUpdater.<init>(ComputationGraphUpdater.java:44)
	at org.deeplearning4j.nn.updater.graph.ComputationGraphUpdater.<init>(ComputationGraphUpdater.java:40)
	at org.deeplearning4j.nn.graph.ComputationGraph.getUpdater(ComputationGraph.java:2807)
	at org.deeplearning4j.nn.graph.ComputationGraph.getUpdater(ComputationGraph.java:2796)
	at org.deeplearning4j.util.ModelSerializer.writeModel(ModelSerializer.java:168)
	at org.deeplearning4j.util.ModelSerializer.writeModel(ModelSerializer.java:118)
	at org.deeplearning4j.util.ModelSerializer.writeModel(ModelSerializer.java:105)
	at org.deeplearning4j.earlystopping.saver.LocalFileGraphSaver.save(LocalFileGraphSaver.java:84)
	at org.deeplearning4j.earlystopping.saver.LocalFileGraphSaver.saveBestModel(LocalFileGraphSaver.java:74)
	at org.deeplearning4j.earlystopping.saver.LocalFileGraphSaver.saveBestModel(LocalFileGraphSaver.java:42)
	at org.deeplearning4j.earlystopping.trainer.BaseEarlyStoppingTrainer.fit(BaseEarlyStoppingTrainer.java:223)
	at anhtv.ai.dl.computervision.SuperImageResolution.train(SuperImageResolution.java:180)
	at anhtv.ai.dl.computervision.SuperImageResolution.main(SuperImageResolution.java:189)


Seem the topologicalOrder is null.

**Comment by tvanh** [comment]
I found the bug in data pipeline. 
java.lang.NullPointerException: null
	at org.deeplearning4j.nn.graph.ComputationGraph.numParams(ComputationGraph.java:3148)
	at org.deeplearning4j.nn.graph.ComputationGraph.numParams(ComputationGraph.java:3142)
	at org.deeplearning4j.nn.graph.ComputationGraph.fitHelper(ComputationGraph.java:1106)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:1098)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:1085)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:965)
	at org.deeplearning4j.earlystopping.trainer.EarlyStoppingGraphTrainer.fit(EarlyStoppingGraphTrainer.java:73)
	at org.deeplearning4j.earlystopping.trainer.BaseEarlyStoppingTrainer.fit(BaseEarlyStoppingTrainer.java:120)
	at anhtv.ai.dl.computervision.SuperImageResolution.train(SuperImageResolution.java:180)
	at anhtv.ai.dl.computervision.SuperImageResolution.main(SuperImageResolution.java:189)

**Comment by tvanh** [comment]
Another NPE. It probly caused by the cropping func. If I want to crop an rect with size (h, w) from image, what should I do?
CropImageTransform crop = new CropImageTransform(0, 0, h, w);
Is the above code correct?

**Comment by saudet** [comment]
No, to get a crop of the size you need try BoxImageTransform:
https://github.com/deeplearning4j/DataVec/blob/master/datavec-data/datavec-data-image/src/main/java/org/datavec/image/transform/BoxImageTransform.java

**Comment by AlexDBlack** [comment]
After spending a while getting your code into something I can actually run (there's a lot of params missing, I had to reverse engineer output size, etc) I'm unable to reproduce this problem.
Here's the code I am using: https://gist.github.com/AlexDBlack/6e0a1dd624becb30ceb00d6f86fb85df
Note the ModelSerializer.writeModel(ModelSerializer.java:105) in your stack trace.

Can you please give me a complete, ready to run code snippet to reproduce this?
I can't help if I'm unable to reproduce the problem.

**Comment by tvanh** [comment]
I fixed the data preprocessing method. However when we tried to train the network using model.fit(data), the network runs well. But if tried to early stop trainer, I got the exception

The code is 

`    public static void main(String[] args) throws Exception {
    	DataSet ds = new DataSet(Nd4j.rand(new int[]{1, 3, 224, 224}), Nd4j.rand(new int[]{1, 1, 238, 238}));
    	DataSetIterator dsi = new ExistingDataSetIterator(ds);
    	ComputationGraph cg = new ComputationGraph(conf());
        cg.init();
        
    	EarlyStoppingModelSaver<ComputationGraph> saver = new LocalFileGraphSaver("D:/data/image/SR_Image");
        EarlyStoppingConfiguration<ComputationGraph> esConf = new EarlyStoppingConfiguration.Builder<ComputationGraph>()
                .epochTerminationConditions(new MaxEpochsTerminationCondition(100), new BestScoreEpochTerminationCondition(0.2)) 
                .evaluateEveryNEpochs(1)
                .iterationTerminationConditions(new MaxTimeIterationTerminationCondition(12, TimeUnit.HOURS)) //Max of 120 minutes
                .scoreCalculator(new DataSetLossCalculator(dsi, true)) 
                .modelSaver(saver)
                .build();
        		
    	EarlyStoppingGraphTrainer trainer = new EarlyStoppingGraphTrainer(esConf, cg, dsi);
        EarlyStoppingResult<ComputationGraph> result = trainer.fit();
        cg = result.getBestModel();
        
        //cg.fit(dsi);
        ModelSerializer.writeModel(cg, new File("deleteme.bin"), true);
    }
`

**Comment by tvanh** [comment]
nvm, I found the bug. The network runs well atm. Waiting for the best model since it will take days to finish.

TYVM for your help.

**Comment by AlexDBlack** [comment]
OK, I'll close this - sounds like this has been resolved.
